staking/contracts/Staking.sol

```diff
@@ -4,87 +4,250 @@ pragma solidity >=0.6.2 <0.8.0;
+        require(amount > 0, "amount negative");
+        require(cliff <= TWO_YEAR_WEEKS, "cliff too big");
+        require(amount.div(slope) <= TWO_YEAR_WEEKS, "period too big");
+        require(token.transferFrom(account, address(this), amount), "transfer failed");
 
-        totalSupplyLine.add(line, cliff);
-        locks[account].balance.add(line, cliff);
-        locks[account].locked.add(lineLocked, cliff);
+        counter++;
+
+        uint blockTime = roundTimestamp(block.timestamp);
+        (uint stAmount, uint stSlope) = getStake(amount, slope, cliff);
+        LibBrokenLine.Line memory line = LibBrokenLine.Line(blockTime, stAmount, stSlope);
+        totalSupplyLine.add(counter, line, cliff);
+        locks[delegate].balance.add(counter, line, cliff);
+        line = LibBrokenLine.Line(blockTime, amount, slope);
+        locks[account].locked.add(counter, line, cliff);
+        deposits[counter].locker = account;
+        deposits[counter].delegate = delegate;
         locks[account].amount = locks[account].amount.add(amount);
-        require(token.transferFrom(account, address(this), amount), "failure while transferring");
+        return counter;
+    }
 
-        // как меняется lock общий, когда юзер приходит/уходит/меняет
-        // 1. нужно применить пропущенные изменения (окончания локов)
-        // 2. если добавляем, то
+    function reStake(uint id, address newDelegate, uint newAmount, uint newSlope, uint newCliff) external notStopped returns (uint) {
+        address account = deposits[id].locker;
+        address delegate = deposits[id].delegate;
+        uint blockTime = roundTimestamp(block.timestamp);
+        verification(account, id, newAmount, newSlope, newCliff, blockTime);
+        removeLines(id, account, delegate, newAmount, blockTime);
+        return addLines(account, newDelegate, newAmount, newSlope, newCliff, blockTime);
     }
 
-    function totalSupply() public returns (uint) {
-        if (totalSupplyLine.initial.start == 0) { //no lock
+    function withdraw() external {
+        uint blockTime = roundTimestamp(block.timestamp);
+        locks[msg.sender].locked.update(blockTime);
+        uint value = locks[msg.sender].amount;
+        if (!stopped) {
+            uint bias = locks[msg.sender].locked.initial.bias;
+            value = value.sub(bias);
+        }
+        if (value > 0) {
+            locks[msg.sender].amount = locks[msg.sender].amount.sub(value);
+            require(token.transfer(msg.sender, value), "Failure while transferring, withdraw");
+        }
+    }
+
+    function delegate(uint id, address newDelegate) external notStopped {
+        address from = deposits[id].delegate;
+        require(from != address(0), "Delegate from address by id not found");
+        LibBrokenLine.LineData memory lineData = locks[from].balance.initiatedLines[id];
+        require(lineData.line.bias != 0, "Line already finished nothing to delegate");
```

Comment:
message is not good. user doesn't know anything about Lines

Reply:
better: deposit already finished
agree? do